### PR TITLE
Mols matrix to grid image

### PR DESCRIPTION
**Pull request by bertiewooster**
#### Reference Issue

Adds functionality #5917

#### What does this implement/fix? Explain your changes.
Adds the functionality to use a two-dimensional (nested) data structure as input to create molecular grid images. For example, the following molecular grid could be created by supplying a nested data structure where
- Each data substructure represents a row
- The length of each data substructure can be different and the RDKit will automatically add empty cells as needed so that all rows are padded to the length of the longest row

Thus, the user can provide a "ragged" data structure (where the length of each data substructure can be different), and the number of columns in the molecular grid image will automatically be set correctly.

Annotated grid of maximum common substructure and core; molecules and groups off maximum common substructure

### CODE REVIEW

**Comment by bertiewooster** [comment]
I'd like to create parametrized tests to cover a variety of situations. Is there a standard way to parametrize unittests? The three options that I found are:

1. [define a function that runs the desired tests, then call it for each parameter set](https://stackoverflow.com/questions/50514539/how-to-parameterize-unittest-test-methods)
2. [ddt](https://github.com/datadriventests/ddt) (Data-Driven Tests, not the chemical ;) whose last commit was Aug 2022
3. [parameterized](https://github.com/wolever/parameterized) which has no commits since Jan 2021 so that suggests it's not maintained anymore

I didn't find any of those options in the RDKit codebase. Absent any feedback here, I guess I'll go with 1. to avoid adding a dependency to the RDKit codebase, especially one that may not be that well maintained.

P.S. Python has a built-in [subTest()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest) context manager but that seems slightly different, not designed to take a range of parameters.

**Comment by greglandrum** [comment]
@bertiewooster : this is still marked as a draft, so I haven't looked at it. Do you intend to come back to it at some point?

**Comment by bertiewooster** [comment]
@greglandrum: no need to review it yet. Yes, I intend to come back to it. I was making good progress then my C got confused about which chip architecture my computer has (`mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')` and I haven't gotten around to fixing it but should have time in the next few weeks. Alternatively I may try the approach in your blog post [Setting up an environment to make Python contributions to the RDKit](https://greglandrum.github.io/rdkit-blog/posts/2020-03-30-setting-up-an-environment.html) because my contribution will be in Python.

**Comment by bertiewooster** [comment]
Usage examples from MolsMatrixToGridImage docstring:

```
        from rdkit import Chem
        from rdkit.Chem.Draw import MolsMatrixToGridImage, rdMolDraw2D
        salt = Chem.MolFromSmiles("[Na][Cl]")
        mols_matrix = [[salt, salt], [salt, None, salt]]

        # Minimal example: Only mols_matrix is supplied,
        # result will be a drawing containing (where each row contains molecules):
        # Cl-Na Cl-Na
        # Cl-Na          Cl-Na
        img = MolsMatrixToGridImage(mols_matrix)
        img.save("MolsMatrixToGridImage_minimal.png")
        # img is a PIL object for a PNG image file like:
        # <PIL.PngImagePlugin.PngImageFile image mode=RGB size=600x200 at 0x1648CC390>
        # Drawing will be saved as PNG file MolsMatrixToGridImage_minimal.png

        # Exhaustive example: All parameters are supplied,
        # result will be a drawing containing (where each row of molecules is followed by a row of legends):
        # 1 Cl-Na 0                          1 Cl-Na 0
        # no highlighting                bond highlighted         
        # 1 Cl-Na 0                                                                   1 Cl-Na 0
        # sodium highlighted                                                  chloride and bond highlighted
        legends_matrix = [["no highlighting", "bond highlighted"], ["sodium highlighted", "", "chloride and bond highlighted"]]
        highlightAtomLists_matrix = [[[],[]], [[0], None, [1]]]
        highlightBondLists_matrix = [[[],[0]], [[], None, [0]]]

        dopts = rdMolDraw2D.MolDrawOptions()
        dopts.addAtomIndices = True

        img_file = MolsMatrixToGridImage(mols_matrix=mols_matrix, subImgSize=(300, 400), legends_matrix=legends_matrix, highlightAtomLists_matrix=highlightAtomLists_matrix, highlightBondLists_matrix=highlightBondLists_matrix, useSVG=False, returnPNG=True, drawOptions=dopts)
        img_file.save("MolsMatrixToGridImage_exhaustive.png")
        # Drawing will be saved as PNG file MolsMatrixToGridImage_exhaustive.png
```

File: rdkit/Chem/Draw/UnitTestDraw.py [hunk 1]
```diff
@@ -7,18 +7,15 @@
...
 import sys
 import tempfile
 import unittest
 
 from rdkit import Chem
-from rdkit.Chem import AllChem
-from rdkit.Chem import Draw
-from rdkit.Chem import rdDepictor
-from rdkit.Chem import rdMolDescriptors
+from rdkit.Chem import AllChem, Draw, rdDepictor, rdMolDescriptors
```
Comment (greglandrum) [inline rdkit/Chem/Draw/UnitTestDraw.py:18]: This change is not connected to this PR and violates the RDKit's normal coding style (it's also incompatible with PEP8)

File: rdkit/Chem/Draw/UnitTestDraw.py [hunk 2]
```diff
@@ -50,6 +47,51 @@ def setUp(self):
...
+    self.highlightAtomLists_matrix = [[ith_item_list(count, natoms, 0) for count in row] for row in repeats]
+
+    # Another bond is created when molecule is oligomerized, so to keep the bond type consistent,
+    #   make items per unit one more than the number of bonds
+    self.highlightBondLists_matrix = [[ith_item_list(count, nbonds + 1, 1) for count in row] for row in repeats]
+
+    # Parametrize tests: In addition to supplying mols_matrix, supply 0-3 other matrices
+    # col labels: legends_matrix,       highlightAtomLists_matrix,      highlightBondLists_matrix      
+    #             Zero other matrices: 1 parameter set
+    self.param_sets = [(None,                None,                           None                          ),
```
Comment (greglandrum) [inline rdkit/Chem/Draw/UnitTestDraw.py:81]: I see what you're trying to do here, but don't think it actually improves readability, please format the code "normally" (i.e. just run `yapf` over it)

File: rdkit/Chem/Draw/__init__.py [hunk 3]
```diff
@@ -14,13 +14,11 @@
 from io import BytesIO
 
 import numpy
-from rdkit import Chem
-from rdkit import RDConfig
-from rdkit import rdBase
+
+from rdkit import Chem, RDConfig, rdBase
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:18]: again, please undo these changes.

File: rdkit/Chem/Draw/__init__.py [hunk 4]
```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
...
+            (in which the molecules and legends are non-nested lists, 
+            and the highlight parameters are two-deep nested lists) 
+
+      RETURNS:
+
+        A grid of molecular images in one of these formats:
+        
+        - useSVG=False and returnPNG=False (default): A PIL object for a PNG image file
+
+        - useSVG=False and returnPNG=True: A PNG file
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:760]: it returns PNG data, not a file

File: rdkit/Chem/Draw/__init__.py [hunk 5]
```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
...
+        highlightAtomLists_matrix = [[[],[]], [[0], None, [1]]]
+        highlightBondLists_matrix = [[[],[0]], [[], None, [0]]]
+
+        dopts = rdMolDraw2D.MolDrawOptions()
+        dopts.addAtomIndices = True
+
+        img_file = MolsMatrixToGridImage(mols_matrix=mols_matrix, subImgSize=(300, 400), 
+        legends_matrix=legends_matrix, highlightAtomLists_matrix=highlightAtomLists_matrix, 
+        highlightBondLists_matrix=highlightBondLists_matrix, useSVG=False, returnPNG=True, drawOptions=dopts)
+        img_file.save("MolsMatrixToGridImage_exhaustive.png")
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:798]: I believe that this is wrong. I think that you should get back a string (binary string, but still a string) when `returnPNG = True`.
You get the PIL object when `returnPNG = False`

File: rdkit/Chem/Draw/__init__.py [hunk 6]
```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
...
+        - highlightAtomLists_matrix: A three-deep nested data structure of integers of atoms to highlight,
+         for example list of lists of lists of integers (default None)
+
+        - highlightBondLists_matrix: A three-deep nested data structure of integers of bonds to highlight,
+         for example list of lists of lists of integers (default None)
+
+        - useSVG: Whether to return an SVG (if true) or PNG (if false);
+         passed through to MolsToGridImage (default false)
+
+        - returnPNG: Whether to return a PNG file (if true) or a PIL object for a PNG image file (if false);
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:733]: when returnPNG is true it returns PNG data, not a file

File: rdkit/Chem/Draw/__init__.py [hunk 7]
```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
...
+
+        - legends: A non-nested data structure of strings to label molecules with, for example list of strings
+
+        - highlightAtomLists: A two-deep nested data structure of integers of atoms to highlight, for example list of lists of integers
+
+        - highlightBondLists: A two-deep nested data structure of integers of bonds to highlight, for example list of lists of integers
+
+  """
+  # Check that each item in nested lists is a list
+  def check_elements_are_lists(nested_list, nested_list_name=""):
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:608]: What's the value of this check? Shouldn't I also be able to provide a list of tuples, or tuple of lists, or tuple of tuples,  etc.? This is Python, there is *no* reason to enforce rigid typing.

I think the requirement here is that you need sequences which support the `len()` function and iteration

Comment (greglandrum) [reply rdkit/Chem/Draw/__init__.py:608]: Second comment here: all of these nested functions make it much harder to actually find the code for this function. Please pull them all out and put them at the global scope with names starting with `_` that are compatible with RDKit naming conventions.

File: rdkit/Chem/Draw/__init__.py [hunk 8]
```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
   return res
 
 
+def _MolsNestedToLinear(mols_matrix, legends_matrix = None, highlightAtomLists_matrix = None,
```
Comment (greglandrum) [inline rdkit/Chem/Draw/__init__.py:574]: General comment on naming: the RDKit generally doesn't use underscores to separate words in variable names (there are instances that slipped through, but they are the minority). Please try to make the code style consistent with the rest of the RDKit.

File: rdkit/Chem/Draw/UnitTestDraw.py [hunk 9]
```diff
@@ -213,6 +255,122 @@ def testGridSVG(self):
...
+
+  def testMolsMatrixToGridImage(self):
+    # Tests are that running Draw.MolsMatrixToGridImage doesn't give an error
+    # Parametrize tests: In addition to supplying mols_matrix, supply 0-3 other matrices
+    subImgSize = (200, 200)
+
+    kwargs_value = Draw.rdMolDraw2D.MolDrawOptions()
+    kwargs_value.addAtomIndices = True
+
+    for (legends_matrix, highlightAtomLists_matrix, highlightBondLists_matrix) in self.param_sets:
```
Comment (greglandrum) [inline rdkit/Chem/Draw/UnitTestDraw.py:366]: Just a comment: these loops aren't really testing anything other than "calling the function in this way doesn't produce an error". Testing drawing code is really hard, and this is a lot better than nothing.
I don't think you need to change anything, but it might be worth a comment at the top of the loop about this

**Review by greglandrum — CHANGES_REQUESTED** [review]
Thanks for the contribution @bertiewooster . I do think this will be useful functionality.
I went through and made a bunch of mainly cosmetic/style suggestions which will make it easier for me to actually review the substance of what's here.
I'll do the real review after that stuff is resolved.

File: rdkit/Chem/Draw/UnitTestDraw.py (hunk 1 above)
Comment (bertiewooster) [reply rdkit/Chem/Draw/UnitTestDraw.py:18]: Oops, Black autoformatter did that despite my trying to prevent it.

File: rdkit/Chem/Draw/__init__.py (hunk 4 above)
Comment (bertiewooster) [reply rdkit/Chem/Draw/__init__.py:760]: Thanks, these return types were difficult for me to determine

File: rdkit/Chem/Draw/__init__.py (hunk 7 above)
Comment (bertiewooster) [reply rdkit/Chem/Draw/__init__.py:608]: First comment: Great point, my intent was to verify that the elements are iterables, not specifically lists. I just changed to check that each element has the `__iter__` and `__len__` attributes.

Second comment: Will do; I debated doing that.

File: rdkit/Chem/Draw/__init__.py (hunk 5 above)
Comment (greglandrum) [reply rdkit/Chem/Draw/__init__.py]: Maybe I'm missing something. The example you just shared has `returnPNG=False`. As I would expect, that returns a PIL object. 
My comment above is about when you do `returnPNG=True`; in that case you should get a string back, but the code I was commenting on was treating it as a PIL object.

Comment (bertiewooster) [reply rdkit/Chem/Draw/__init__.py]: My mistake, I thought I was using the same options in my code as in the docstring but I must've changed them sometime along the line. I'll revisit.

File: rdkit/Chem/Draw/__init__.py (hunk 7 above)
Comment (greglandrum) [reply rdkit/Chem/Draw/__init__.py]: > First comment: Great point, my intent was to verify that the elements are iterables, not specifically lists. I just changed to check that each element has the `__iter__` and `__len__` attributes.

But you don't need to do that explicitly since Python will do it for you and raise a TypeError with an informative message if the caller gets it wrong:
```
In [32]: def loopy(thing):
    ...:     res = [0]*len(thing)
    ...:     for i,item in enumerate(thing):
    ...:         res[i] = item+1
    ...:     return res
    ...: 

In [33]: loopy((1,2,3))
Out[33]: [2, 3, 4]

In [34]: loopy(123)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [34], line 1
----> 1 loopy(123)

Cell In [32], line 2, in loopy(thing)
      1 def loopy(thing):
----> 2     res = [0]*len(thing)
      3     for i,item in enumerate(thing):
      4         res[i] = item+1

TypeError: object of type 'int' has no len()
```

There's no sense in doing this check yourself, it's much better to let the language do the work for you and ensure that the same exceptions are thrown for the same types of errors.

**Comment by bertiewooster** [comment]
Hi, I'm re-requesting substantive review because I believe I addressed all @greglandrum's cosmetic/style suggestions above. Thanks!

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
Thanks @bertiewooster !

**Comment by bertiewooster** [comment]
Great, thanks @greglandrum! I'm excited to have made a functionality contribution to the RDKit.

Regarding documentation for this new feature,

- Because this is a graphical feature, a visual format such as a blog post would be helpful for explaining it. I'll work on a Jupyter Notebook and would be happy to contribute it to the [RDKit blog](https://greglandrum.github.io/rdkit-blog/) if you like.
- I assume the API documentation such as function signature (similar to that of [MolsToGridImage](https://www.rdkit.org/docs/source/rdkit.Chem.Draw.html#rdkit.Chem.Draw.MolsToGridImage)) will be automatically generated from the [docstring in the code](https://github.com/rdkit/rdkit/blob/master/rdkit/Chem/Draw/__init__.py#L669).

**Comment by greglandrum** [comment]
> 
> * Because this is a graphical feature, a visual format such as a blog post would be helpful for explaining it. I'll work on a Jupyter Notebook and would be happy to contribute it to the [RDKit blog](https://greglandrum.github.io/rdkit-blog/) if you like.

That would be great!

> * I assume the API documentation such as function signature (similar to that of [MolsToGridImage](https://www.rdkit.org/docs/source/rdkit.Chem.Draw.html#rdkit.Chem.Draw.MolsToGridImage)) will be automatically generated from the [docstring in the code](https://github.com/rdkit/rdkit/blob/master/rdkit/Chem/Draw/__init__.py#L669).

Yep, that should happen automatically.

**Comment by bertiewooster** [comment]
I've got the Jupyter Notebook for explaining `MolsMatrixToGridImage()` ready and am trying to run it with the [RDKit 2023_09_1 release](https://github.com/rdkit/rdkit/releases/tag/Release_2023_09_1); I am having trouble upgrading to that version on my Mac. When I follow the [installation instructions](http://rdkit.org/docs/Install.html#how-to-install-rdkit-with-conda) 

```
conda create -c conda-forge -n my-rdkit-env rdkit` 
conda activate my-rdkit-env
```

the RDKit version is 2022.03.5:

```
# Name                    Version                   Build  Channel
rdkit                     2022.03.5       py310h9a91a65_0    conda-forge
```

Running `conda upgrade rdkit` does not upgrade the RDKit version. Is there some other way to get 2023_09_1? I checked PyPI for the `pip`-installable version, but [PyPI is on version 2023.3.3](https://pypi.org/project/rdkit/).

**Comment by greglandrum** [comment]
Maybe try `conda install rdkit=2023.09.1`?

**Comment by bertiewooster** [comment]
Thanks for the tip! What ended up working was 

```
conda install -c conda-forge rdkit=2023.09.1
```

I initially tried your suggestion as-is and got 

```
PackagesNotFoundError: The following packages are not available from current channels:

  - rdkit=2023.09.1

Current channels:

  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch
```

so I gathered that part of the problem was the latest RDKit wasn't in the Anaconda channels.

**Comment by greglandrum** [comment]
> so I gathered that part of the problem was the latest RDKit wasn't in the Anaconda channels

To the best of my knowledge no version of the RDKit is in the Anaconda channels. It's always been in conda-forge.

**Comment by bertiewooster** [comment]
@greglandrum I completed drafting the blog post and submitted a [pull request to the RDKit Blog repo](https://github.com/greglandrum/rdkit_blog/pull/16).
